### PR TITLE
Remove crypto dependency on secvarctl's generic functions

### PR DESCRIPTION
**Pull request by nick-child-ibm**
This pull request edits the `crypto` functions to not depend on secvarctls `generic` functions. Functions defined in `generic.c` are ease-of-use functions that simplify basic I/O tasks. Previously, functions in `crypto` used these `generic` functions to save some effort. With the recent goal to move these `crypto` functions into skiboot, we must remove this dependency. This PR removes all calls from the crypto functions to the generic functions.

### CODE REVIEW

File: external/skiboot/libstb/secvar/crypto/crypto-openssl.c
Comment (daxtens) [inline external/skiboot/libstb/secvar/crypto/crypto-openssl.c:669]: (I realise that you're copying over your existing behaviour, but...) Wouldn't it be more conventional to use `:` as a separator rather than `/`?

Also, perhaps to avoid the separator preceding the first hash value.
```
printf("%02x", (*outHash)[0]);
for (int i = 1; i < *outHashSize; i++) // (also, does our C spec variant allow loop counters to be declared here?)
    printf(":%02x", ...)
```

File: external/skiboot/libstb/secvar/crypto/crypto-openssl.c
Comment (daxtens) [inline external/skiboot/libstb/secvar/crypto/crypto-openssl.c:285]: Hmm, that's a slightly concerning cast. We should probably make crtSize (and keySize) a long and then cast them to the required type - possibly with a test or assert to make sure they're non-negative. Harder to be stuffed around with integer casts than pointer casts, IMO.

File: external/skiboot/libstb/secvar/crypto/crypto-mbedtls.c
Comment (daxtens) [inline external/skiboot/libstb/secvar/crypto/crypto-mbedtls.c:17]: Yeah this is a good start. I wonder if we should wait until you have all the parts to get this to compile purely as part of skiboot and then go from there?

**Review by daxtens — CHANGES_REQUESTED** [review]
This is a really good start, I have only minor nits.

I wonder if we should try and bundle together all the changes that are required to get the crypto abstraction layer to compile as part of skiboot, rather than go piecewise into secvarctl? Not wildly fussy though, if you or @erichte-ibm have strong feelings.

**Comment by nick-child-ibm** [comment]
> I wonder if we should try and bundle together all the changes that are required to get the crypto abstraction layer to compile as part of skiboot, rather than go piecewise into secvarctl? Not wildly fussy though, if you or @erichte-ibm have strong feelings.

I think that is a good idea. When I am done getting this crypto stuff to work on SkiBoot, I will just copy it into secvarctl and commit.

File: external/skiboot/libstb/secvar/crypto/crypto-openssl.c
Comment (erichte-ibm) [reply external/skiboot/libstb/secvar/crypto/crypto-openssl.c:669]: We're not adhering to a particularly restrictive C spec (that I'm aware of), but I would prefer not having initializations inside for-loops.

**Comment by erichte-ibm** [comment]
> > I wonder if we should try and bundle together all the changes that are required to get the crypto abstraction layer to compile as part of skiboot, rather than go piecewise into secvarctl? Not wildly fussy though, if you or @erichte-ibm have strong feelings.
> 
> I think that is a good idea. When I am done getting this crypto stuff to work on SkiBoot, I will just copy it into secvarctl and commit.

Thirding this. I don't think it's a bad idea to keep the crypto abstraction in skiboot, in case we decide to switch to openssl (or some other library) in skiboot for some crazy reason. Ideally if `libstb` is to stand more on its own, then it'd make sense for it to be usable as part of skiboot OR secvarctl without any modification.

File: external/skiboot/libstb/secvar/crypto/crypto-openssl.c
Comment (nick-child-ibm) [reply external/skiboot/libstb/secvar/crypto/crypto-openssl.c:669]: This may be worthy of an entire new PR. All for loops in secvarctl use internal counter initializations. In terms of C spec, we use gnu99 which is C99 with gnu extensions. It appears that C99 allows for declarations in a for loop. Let me know where we want to go with this.

Comment (erichte-ibm) [reply external/skiboot/libstb/secvar/crypto/crypto-openssl.c:669]: Considering this is a skiboot file, we should adhere to their usage. A quick grep shows at least a few instances of initializing in the loop, but ultimately it will be up to the current maintainer.

I value consistency more than my personal gripes though, so if these are already consistently used in secvarctl, I'm fine with whichever.

**Comment by nick-child-ibm** [comment]
I should have possible made another pull request to redefine its scope but I felt lazy so hopefully this will do:
The new commits were all part of the effort to compile a skiboot branch with the crypto abstraction functions. This has been done [here](https://github.com/nick-child-ibm/skiboot/tree/crypto_devel). Once I got skiboot to compile, with some edits to the `crypto` files, I have now copied those files back into secvarctl so that we are one step closer to submoduling. 

Some interesting things came up in this porting-a-port process. Mainly, we now have 3 versions of both `pkcs7_write.c` and `pkcs7.c`. They both exist in skiboot, mbedtls PR's and secvarctl. I believe, as of this PR, all of the `pkcs7_write.c` are equivalent. But the `pkcs7.c` files all contain different functionalities and return codes. We should look to update these soon. Additionally, while adding `crypto` to skiboot, I realized we don't need any of the `crypto_write...` functions and therefore we do not need `pkcs7_write.c` in skiboot. I added it to the source tree but do not compile it.

**Comment by erichte-ibm** [comment]
>  Additionally, while adding crypto to skiboot, I realized we don't need any of the crypto_write... functions and therefore we do not need pkcs7_write.c in skiboot. I added it to the source tree but do not compile it.

Hmm, the main focus for `NO_CRYPTO` was probably for skiroot (not skiboot). Is there value in still having an actually fully `NO_CRYPTO`'d build? How much functionality is lost? Obviously we can't generate updates, but can we still enroll and display variables?

Either way, we shouldn't need to include `pkcs7_write.c` at all in skiboot, I'd `#ifdef` out any references to those functions in the other secvar/ctl code.

**Comment by nick-child-ibm** [comment]
> Is there value in still having an actually fully `NO_CRYPTO`'d build?
Because the macro name `NO_CRYPTO` implies there being no crypto functions, I have changed it to `CRYPTO_READ_ONLY`.  This is mainly because crypto library functions are still required for reading and updating secvars (things like `mbedtls_x509_parse`). I know this isn't exactly "crypto" but I think it makes more sense as `CRYPTO_READ_ONLY`. At this moment, `CRYPTO_READ_ONLY` completely `#idfef`'s out all of `secvarctl generate` related code as well as all of the `crypto` functions that are specific to `secvarctl generate` . There was one final `#ifdef`ing that occurs is `edk2-svc-validate.c` because it was using a  function in `pkcs7_write.c` (if `CRYPTO_READ_ONLY` then certificates can only be parsed in DER, input certs in PEM are not accepted). I think this makes sense, it has been some time since I examined the difference in executable size but I am not sure what else to remove. 


> Either way, we shouldn't need to include `pkcs7_write.c` at all in skiboot, I'd `#ifdef` out any references to those functions in the other secvar/ctl code.
I have removed pkcs7_write.c and reverted any changes made to `libstb/crypto/pkcs7/` . Rebased and tested everything works fine with preprocessor directives factoring out unnecessary functions.

**Comment by nick-child-ibm** [comment]
Rebased to match `crypto` files with most recent of those in my skiboot branch. The changes are limited to adding a preprocessor directive around `#include pkcs7_write.h` and fixing some whitespace in source code.

**Comment by nick-child-ibm** [comment]
Rebased onto new `main`

File: CMakeLists.txt
Comment (erichte-ibm) [inline CMakeLists.txt:71]: I like this a lot better. :+1: 

Side question: it doesn't matter whether we use either crypto lib for this right?
Side side question: does this work for both static and dynamic linked builds?
Side side side question: can we build static with all crypto libs we support, or is that limited to mbedtls only?

None of the above questions are relevant to this PR, just wondering for future.

File: external/skiboot/libstb/secvar/crypto/crypto-mbedtls.c
Comment (erichte-ibm) [inline external/skiboot/libstb/secvar/crypto/crypto-mbedtls.c:27]: Modification of code in `external`, is this necessary, or is this something we can adjust upstream as well? I know this is a bit of an old PR and I'm not up to date on where things are at the moment.

**Comment by erichte-ibm** [comment]
Mostly looks fine to me, going to defer to @daxtens 's judgement to merge as he still has the open changes requested.

File: external/skiboot/libstb/secvar/crypto/crypto-mbedtls.c
Comment (erichte-ibm) [reply external/skiboot/libstb/secvar/crypto/crypto-mbedtls.c:27]: The plan is to eventually have this code identical between skiboot and secvarctl, correct? If that is the case, then these paths should be unified between the two (somehow) so we don't need to modify skiboot code to work with secvarctl.

If there's ongoing effort elsewhere to fix this (possibly what Daniel is working on with libstb-secvar), then we can overlook this for now.

**Pull request merged**
